File: src/main/java/pl/mjedynak/idea/plugins/builder/action/handler/GoToBuilderActionHandler.java
package pl.mjedynak.idea.plugins.builder.action.handler;

import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.editor.Editor;
import com.intellij.psi.PsiClass;
import javax.swing.JList;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import pl.mjedynak.idea.plugins.builder.factory.GoToBuilderPopupListFactory;
import pl.mjedynak.idea.plugins.builder.gui.displayer.GoToBuilderPopupDisplayer;
import pl.mjedynak.idea.plugins.builder.psi.PsiHelper;

public class GoToBuilderActionHandler extends AbstractBuilderActionHandler {

    public GoToBuilderActionHandler(
            GoToBuilderPopupDisplayer popupDisplayer,
            GoToBuilderPopupListFactory popupListFactory,
            DisplayChoosers displayChoosersRunnable) {
        super(popupDisplayer, popupListFactory, displayChoosersRunnable);
    }

    @Override
    protected void doActionWhenClassToGoIsFound(
            @NotNull Editor editor,
            @NotNull PsiClass psiClassFromEditor,
            @Nullable DataContext dataContext,
            boolean isBuilder,
            @NotNull PsiClass classToGo) {
        PsiHelper.navigateToClass(classToGo);
    }

    @Override
    protected void doActionWhenClassToGoIsNotFound(
            @NotNull Editor editor,
            @NotNull PsiClass psiClassFromEditor,
            @Nullable DataContext dataContext,
            boolean isBuilder) {
        if (!isBuilder) {
            displayPopup(editor);
        }
    }

    private void displayPopup(@NotNull Editor editor) {
        JList<?> popupList = popupListFactory.getPopupList();
        popupDisplayer.displayPopupChooser(editor, popupList, () -> displayChoosers.run(null));
    }
}

File: src/main/kotlin/io/github/kingg22/kobra/builder/action/handler/GoToBuilderActionHandler.kt
package io.github.kingg22.kobra.builder.action.handler

import com.intellij.openapi.actionSystem.DataContext
import com.intellij.openapi.editor.Editor
import com.intellij.psi.PsiClass
import io.github.kingg22.kobra.builder.factory.GoToBuilderPopupListFactory
import io.github.kingg22.kobra.builder.gui.displayer.GoToBuilderPopupDisplayer
import io.github.kingg22.kobra.builder.psi.PsiHelper
import org.jetbrains.annotations.VisibleForTesting

public class GoToBuilderActionHandler : AbstractBuilderActionHandler {
    public constructor() : super(GoToBuilderPopupDisplayer(), GoToBuilderPopupListFactory())

    @VisibleForTesting
    internal constructor(
        popupDisplayer: GoToBuilderPopupDisplayer,
        popupListFactory: GoToBuilderPopupListFactory,
        displayChoosers: DisplayChoosers,
    ) : super(popupDisplayer, popupListFactory, displayChoosers)

    override fun doActionWhenClassToGoIsFound(
        editor: Editor,
        psiClassFromEditor: PsiClass,
        dataContext: DataContext?,
        isBuilder: Boolean,
        classToGo: PsiClass,
    ) {
        PsiHelper.navigateToClass(classToGo)
    }

    override fun doActionWhenClassToGoIsNotFound(
        editor: Editor,
        psiClassFromEditor: PsiClass,
        dataContext: DataContext?,
        isBuilder: Boolean,
    ) {
        if (!isBuilder) {
            displayPopup(editor)
        }
    }

    private fun displayPopup(editor: Editor) {
        val popupList = popupListFactory.popupList
        popupDisplayer.displayPopupChooser(editor, popupList) { displayChoosers.run(null) }
    }
}
